osd: add duration field to dump_historic_ops method

ideepika opened the pull request:
To make dump_historic_ops more human readable, add duration along
with event and stamp fields to dump_historic_ops method.

Signed-off-by: Deepika Upadhyay <deepikaupadhyay01@gmail.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

markhpc commented:
Hi,

I like the idea of the changes, but I would like to make sure that we aren't introducing any performance regressions here as the OpTracker is fairly hot during heavy small IO.  Have you done any profiling or performance tests with your changes?

ideepika commented:
> Hi,
> 
> I like the idea of the changes, but I would like to make sure that we aren't introducing any performance regressions here as the OpTracker is fairly hot during heavy small IO. Have you done any profiling or performance tests with your changes?
Hey! :)
Nope. Can you tell me some reference, I could learn to do performance tests for Ceph from and test these changes?

tchaikov commented on src/common/TrackedOp.h:
```diff
@@ -190,19 +190,35 @@ class OpTracker {
+	MarkEventStore(std::string s, utime_t t) : str(s), stamp(t) {}
+      };
+
+      auto compare = [] (auto first, auto second) -> bool {
```
i am confused. why do we need to sort by the timestamp? as i think the events are always triggered in chronological order. like

1. header_read
2. throttled
3. all_read
4. dispatched

markhpc replied:
Agreed.  Also, we really don't want to sort here in any event for the same reasons as above.

ideepika replied:
In the dump_historic_ops, we got the resulting stamp for throttled is earlier to header_read; and then on later test, also to initiated timestamp; Josh suggested as they are coming straight from messenger, this can happen.
`    [
                        "initiated",
                        "2019-06-28T16:22:45.312798-0400",
                        0
                    ],
                    [
                        "header_read",
                        "2019-06-28T16:22:45.312798-0400",
                        4294967295.9938254
                    ],
                    [
                        "throttled",
                        "2019-06-28T16:22:45.306623-0400",
                        0.0062214990000000001
                    ],
                    [
                        "all_read",
                        "2019-06-28T16:22:45.312844-0400",
                        1.1671000000000001e-05
                    ],
  `

ideepika replied:
I always got throttled stamp earlier than header_read, will updating their order alone would be better as per keeping optimization in mind?

ideepika commented:
> Looks good to me! Nice commit message explaining why changes were done, too!

Thanks @jdurgin!

tchaikov commented on src/common/TrackedOp.h:
```diff
@@ -190,19 +190,35 @@ class OpTracker {
-      retval->mark_event("throttled", params->get_throttle_stamp());
-      retval->mark_event("all_read", params->get_recv_complete_stamp());
-      retval->mark_event("dispatched", params->get_dispatch_stamp());
+      struct MarkEventStore {
```
i'd recommend: s/MarkEventStore/mark_event_t/

ideepika replied:
yeah, would be better!

markhpc commented:
> > Hey! :)
> > Nope. Can you tell me some reference, I could learn to do performance tests for Ceph from and test these changes?

Probably the best way would be to run fio with the librbd engine against a single OSD on a very fast storage device like NVMe.  Something like this should work:

```
fio --ioengine=rbd --direct=1 --bs=4096B --iodepth=128 --end_fsync=0 --rw=randwrite --norandommap --size=262144M --numjobs=1 --runtime=1200 --time_based --ramp_time=None --clientname=admin --pool=cbt-librbd --rbdname=`hostname -f`-1 --invalidate=0 --name=cbt-librbd/`hostname -f`-1-0
```

Otherwise I can try to give it a quick spin layered on top of my optimization PR once I've got it finalized.

Mark

ideepika commented on src/common/TrackedOp.h:
```diff
@@ -190,19 +190,35 @@ class OpTracker {
-      retval->mark_event("all_read", params->get_recv_complete_stamp());
-      retval->mark_event("dispatched", params->get_dispatch_stamp());
+      struct MarkEventStore {
+	std::string str;
```
sure!

markhpc replied:
We really don't want any local copies at all as this is a very hot part of the code.

ideepika commented:
I'll update changes and try to make it work without sort then, be right back.

tchaikov commented on src/common/TrackedOp.h:
```diff
@@ -190,19 +190,35 @@ class OpTracker {
+	std::string str;
+	utime_t stamp;
+
+	MarkEventStore(std::string s, utime_t t) : str(s), stamp(t) {}
```
please pass by `string_view`, a temporary instance of `std::string` is expensive.

tchaikov commented on src/common/TrackedOp.h:
```diff
@@ -190,19 +190,35 @@ class OpTracker {
+      return first.stamp < second.stamp;
+      };
+
+      std::vector<MarkEventStore> events = {
```
might want to use `array<>` if the size is known before hand.

ideepika commented:
> > > Hey! :)
> > > Nope. Can you tell me some reference, I could learn to do performance tests for Ceph from and test these changes?
> 
> Probably the best way would be to run fio with the librbd engine against a single OSD on a very fast storage device like NVMe. Something like this should work:
> 
> ```
> fio --ioengine=rbd --direct=1 --bs=4096B --iodepth=128 --end_fsync=0 --rw=randwrite --norandommap --size=262144M --numjobs=1 --runtime=1200 --time_based --ramp_time=None --clientname=admin --pool=cbt-librbd --rbdname=`hostname -f`-1 --invalidate=0 --name=cbt-librbd/`hostname -f`-1-0
> ```
> 
> Otherwise I can try to give it a quick spin layered on top of my optimization PR once I've got it finalized.
> 
> Mark

Hey @markhpc!  
I have updated my PR, and just changed the throttled event order, as it is the one with earlier stamp than  header_read; I don't know the reason as of now, why it's getting recorded earlier, but this rearrangement fixes the wrong duration(negative integer) issue for the duration. 
It would be nice if you can along with your PR, also open to suggestions on any better way I can fix this

ideepika commented:
retest this please

jdurgin commented on src/osd/OpRequest.cc:
```diff
@@ -30,12 +30,15 @@ using std::stringstream;
-  hitset_inserted(false)
-{
+OpRequest::OpRequest(Message* req, OpTracker* tracker)
+    : TrackedOp(tracker, req->get_recv_stamp() > req->get_throttle_stamp()
```
if the throttle step is always before header_read, you could always use the throttle_stamp here and eliminate the conditional

ideepika replied:
agreed! I'll test it once more to reconfirm and update commit.

ideepika replied:
Yep, can verify the results are satisfactory on replacing `req->get_recv_stamp())` with `req->get_throttle_stamp()`
```
{
                        "event": "initiated",
                        "time": "2019-07-02T12:23:58.513422-0400",
                        "duration": 0
                    },
                    {
                        "event": "throttled",
                        "time": "2019-07-02T12:23:58.513422-0400",
                        "duration": 0.0080613360000000005
                    },
                    {
                        "event": "header_read",
                        "time": "2019-07-02T12:23:58.521483-0400",
                        "duration": 3.5200000000000002e-05
                    },
```

markhpc commented on src/osd/OpRequest.cc:
```diff
@@ -64,11 +65,25 @@ void OpRequest::_dump(Formatter *f) const
+      f->dump_stream("time") << i->stamp;
+
+      auto i_next = i;
+      i_next++;
```
Right now this can increment past the last event.  You are safe below so long as the last thing you see is "done", but it would be better to be explicit about it.  Maybe change this to something like:

auto i_next = i+1;
if (i_next < events.end()) {
  // Deal with the typical case
} else {
  // Deal with the done case and "not done" case when there are no more events
}

ideepika replied:
Sure! yeah, should have considered corner cases, thanks. Updating.

ideepika replied:
can go with
`      if (i_next < events.end()) {
	f->dump_float("duration", i_next->stamp - i->stamp);
      } else if (i->compare("done") == 0) {
	f->dump_float("duration", i->stamp - get_initiated());
      }
`
this way both conditions will be guarded, and even if i_next, exceeds, nothing would be affected.

markhpc commented:
Tested yesterday on top of my optimization branch.  No performance issues!  I did see the same throttled issue but I must have pulled before the latest change.  Other than checking against the end iterator this looks good to me!

tchaikov commented:
- http://pulpito.ceph.com/kchai-2019-07-11_01:43:21-rados-wip-kefu-testing-2019-07-09-1756-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-07-09_13:03:00-rados-wip-kefu-testing-2019-07-09-1756-distro-basic-mira/

test failures are tracked by

- http://tracker.ceph.com/issues/40725
- http://tracker.ceph.com/issues/40618
- http://tracker.ceph.com/issues/40726

gmayyyha commented:
@ideepika mds also has dump_historic_ops.

ideepika commented:
@gmayyyha I am not really much familiar with mds right now, I'll test and try to include changes to update dump_historic_ops for mds too.

neha-ojha commented:
> @ideepika mds also has dump_historic_ops.

@batrick FYI
